rust/src/builtins/scriptlet_intercept/useradd.rs

```diff
@@ -0,0 +1,236 @@
+    // but in the end we only extract the username and UID/GID
+    // (if present).
+    let matches = cli_cmd().get_matches_from(args);
+    let username = matches
+        .value_of("username")
+        .ok_or_else(|| anyhow!("missing required username argument"))?;
+    let uid = matches
+        .value_of("uid")
+        .map(|s| s.parse::<u32>())
+        .transpose()?;
+    let group = matches.value_of("group");
+    let gecos = matches.value_of("comment");
+    let homedir = matches.value_of("homedir");
+    let shell = matches.value_of("shell");
+
+    if !matches.is_present("system") {
+        eprintln!("Trying to create non-system user '{username}'; this will become an error in the future.");
+    }
+
+    let rootdir = Dir::open_ambient_dir("/", cap_std::ambient_authority())?;
+    generate_sysusers_fragment(
+        &rootdir,
+        &pkgname,
+        username,
+        (uid, group),
+        gecos,
+        homedir,
+        shell,
+    )?;
+
+    Ok(())
+}
+
+/// CLI parser, matches <https://linux.die.net/man/8/useradd>.
+fn cli_cmd() -> Command<'static> {
+    let name = "useradd";
+    Command::new(name)
+        .bin_name(name)
+        .about("create a new user")
+        .arg(
+            Arg::new("comment")
+                .short('c')
+                .long("comment")
+                .takes_value(true),
+        )
+        .arg(
+            Arg::new("homedir")
+                .short('d')
+                .long("home-dir")
+                .takes_value(true),
+        )
+        .arg(Arg::new("group").short('g').long("gid").takes_value(true))
+        .arg(Arg::new("system").short('r').long("system"))
+        .arg(Arg::new("shell").short('s').long("shell").takes_value(true))
+        .arg(Arg::new("uid").short('u').long("uid").takes_value(true))
+        .arg(Arg::new("username").required(true))
+}
+
+/// Write a sysusers.d configuration fragment for the given user.
+///
+/// This returns whether a new fragment has been actually written
+/// to disk.
+fn generate_sysusers_fragment(
+    rootdir: &Dir,
+    pkgname: &str,
+    username: &str,
+    id: (Option<u32>, Option<&str>),
+    gecos: Option<&str>,
+    homedir: Option<&str>,
+    shell: Option<&str>,
+) -> Result<bool> {
+    static SYSUSERS_DIR: &str = "usr/lib/sysusers.d";
+
+    // The filename of the configuration fragment is in fact a public
+    // API, because users may have masked it in /etc. Do not change this.
+    let filename = format!("40-pkg-{pkgname}-user-{username}.conf");
+
+    rootdir.create_dir_all(SYSUSERS_DIR)?;
+    let conf_dir = rootdir.open_dir(SYSUSERS_DIR)?;
+    if conf_dir.exists(&filename) {
+        return Ok(false);
+    }
+
+    let id = match id {
+        (Some(uid), Some(group)) => format!("{uid}:{group}"),
+        (Some(id), None) => format!("{id}"),
+        (None, Some(group)) => format!("-:{group}"),
+        (None, None) => "-".to_string(),
+    };
+
+    let gecos = gecos
+        .map(|g| format!("\"{g}\""))
```

Comment:
Perhaps pathological but should we try to handle quotes in here?  Hmm, I'm not seeing documentation around the syntax, e.g. whether it's possible to escape strings etc.

This is definitely a downside of systemd's tendency to create ad-hoc file formats.

Perhaps just to raise awareness of this issue, we could break out a `fn quote_gecos(gecos: &str) -> String` which is perhaps just:

```
assert!(!gecos.contains('"'));
format!(r#""{gecos}""#)
```
or so?

Reply:
Ah, good point. I'll double check the parser in systemd, but I think it supports basic quoting. We already have some shell-quoting logic that we re-use here.